### PR TITLE
Feat: random bundles test

### PR DESCRIPTION
**Pull request by bluekirby1111**
Adds random bundles generation and tests within rust-sdk

### CODE REVIEW

File: src/index.rs [hunk 1]
```diff
@@ -17,20 +17,30 @@ use crate::Secp256k1Signer;
...
+    Solana = 4,
+    InjectedAptos = 5,
+    MultiAptos = 6,
+    Cosmos, //TODO: assign constant
 }
 
 pub struct Config {
     pub sig_length: usize,
     pub pub_length: usize,
+    pub sig_name: String,
```
Comment (mikonieminen) [inline src/index.rs:43]: Rather store the signature type here and then implement `as_str` and maybe you want also `to_string` (that would just call `as_str().to_string()`).

File: src/index.rs [hunk 2]
```diff
@@ -17,20 +17,30 @@ use crate::Secp256k1Signer;
...
+#[cfg(feature = "aptos")]
+use crate::AptosSigner;
+
+#[cfg(feature = "aptos")]
+use crate::MultiAptosSigner;
+
 use crate::error::BundlrError;
 
 #[derive(FromPrimitive, Display, PartialEq, Eq, Debug, Clone)]
 pub enum SignerMap {
```
Comment (mikonieminen) [inline src/index.rs:29]: I wanted to bring this up quite some time so doing it now: this type name is quite bad, it always makes me wonder what is this about until I look at the definition. Maybe this should be renamed to `SignerType` or something else. In the end, it's not a `Map` type, it's an enumeration.

File: src/transaction/bundlr.rs [hunk 3]
```diff
@@ -32,38 +32,46 @@ pub struct BundlrTx {
...
-        }
+        })
     }
 
     fn from_info_bytes(buffer: &[u8]) -> Result<(Self, usize), BundlrError> {
         let sig_type_b = &buffer[0..2];
-        let signature_type = u16::from_le_bytes(<[u8; 2]>::try_from(sig_type_b).unwrap());
+        let signature_type = u16::from_le_bytes(
+            <[u8; 2]>::try_from(sig_type_b)
+                .map_err(|err| BundlrError::BytesError(err.to_string()))?,
```
Comment (mikonieminen) [inline src/transaction/bundlr.rs:59]: I would move this to a separate variable and pass it here. Would make the code a bit cleaner.

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        ],
+        "esModuleInterop": true,
+        "alwaysStrict": true,
+        "sourceMap": true,
+        "strictNullChecks": false,
+        "experimentalDecorators": true,
+        "strict": true,
+        "resolveJsonModule": true
+    },
+}
```
Comment (mikonieminen) [inline tsconfig.json:25]: Missing new line at the end of the file. Please, fix your editor.

File: bundles.ts
Comment (mikonieminen) [inline bundles.ts:1]: Please move all of this nodejs specific code into a separate subdir.

**Review by mikonieminen — CHANGES_REQUESTED** [review]
Some of the changes can be done with separate refactoring, but those should be tracked as separate tickets. If moved to separate ticket, reply in the comment with ticket ID.

File: examples/fund.rs [hunk 5]
```diff
@@ -7,8 +7,10 @@ use reqwest::Url;
 async fn main() {
     let url = Url::parse("https://node1.bundlr.network").unwrap();
     let wallet = PathBuf::from_str("res/test_wallet.json").unwrap();
-    let currency = Arweave::new(wallet, None);
-    let bundlr = Bundlr::new(url, &currency).await;
+    let currency = Arweave::new(wallet, None).expect("Could not create currency");
```
[the remark this answers is not in the document]
Comment (bluekirby1111) [reply examples/fund.rs:10]: Refactoring to build pattern is better approach

[the remark this answers is not in the document]
Comment (bluekirby1111) [reply examples/fund.rs:10]: Naming it as `Winston` (base unit for Arweave) would create further confusion, Arweave is better

[the remark this answers is not in the document]
Comment (bluekirby1111) [reply examples/fund.rs:10]: Some functionalities from Currency needs private key. The main idea was to have this as an option

File: src/index.rs (hunk 1 above)
Comment (bluekirby1111) [reply src/index.rs:43]: I'll mark this as future refactor

File: src/currency/arweave.rs [hunk 6]
```diff
@@ -124,13 +133,13 @@ impl Currency for Arweave {
...
-            .sign(Bytes::copy_from_slice(message))
-            .expect("Could not sign message")
-            .to_vec()
+    fn sign_message(&self, message: &[u8]) -> Result<Vec<u8>, BundlrError> {
+        match &self.signer {
+            Some(signer) => Ok(signer.sign(Bytes::copy_from_slice(message))?.to_vec()),
+            None => Err(BundlrError::CurrencyError(
+                "No private key present".to_string(),
+            )),
+        }
```
[the remark this answers is not in the document]
Comment (bluekirby1111) [reply src/currency/arweave.rs:181]: Currencies have public functions, like verifying or just querying balance. You don't need a signer to use these

File: src/error.rs [hunk 7]
```diff
@@ -62,6 +62,39 @@ pub enum BundlrError {
...
+    ArweaveSdkError(arweave_rs::error::Error),
+
+    #[error("Currency error: {0}")]
+    CurrencyError(String),
+
+    #[error("Error reading/writting bytes: {0}")]
+    BytesError(String),
+
+    #[error("Error converting type: {0}")]
+    TypeParseError(String),
```
[the remark this answers is not in the document]
Comment (bluekirby1111) [reply src/error.rs:75]: Good catch

File: src/index.rs (hunk 1 above)
Comment (bluekirby1111) [reply src/index.rs:43]: These errors will need some further refactoring, also.

File: examples/fund.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikonieminen) [reply examples/fund.rs]: When looking at the `Currency` trait, I can see there are some mistakes in the design (which require you to pass wallet/signer to the `Currency` instance):
1. `get_tx` and `get_tx_status` are really weird ones there, why would you ask information about transactions from a currency?
2. `get_current_height`, what does this to do with a currency?
3. all signing and verifying of transactions, these are signer's (or wallet's) functionalities, not currency's
4. `price`, this is either not something a currency should provide, this is something exchange tells you

Looking at the trait, about 90% of it is in a wrong place.

File: .gitignore [hunk 8]
```diff
@@ -4,3 +4,7 @@ test_item
 wallet.json
 .wallet.json
 test_data_item*
+node_modules
+dist
+/res/gen_bundles/bundle*
+bundles.js
```
Comment (mikonieminen) [inline .gitignore:10]: Missing newline. Fix your editor.

**Comment by bluekirby1111** [comment]
Smaller issues noted, I'll address them soon

**Pull request merged**